app/public/cantusdata/helpers/source_importer.py

```diff
@@ -0,0 +1,124 @@
+        """
+        Requests data on an individual source in CantusDB using
+        the /json-node/ endpoint.
+        
+        :param str source_id: the ID of the source in CantusDB
+        :return: a parsed json object containing source data
+        """
+        source_url = self.cdb_base_url + "/json-node/" + source_id
+        source_response = request.urlopen(source_url).read()
+        source_json = json.loads(source_response)
+        return source_json
+    
+    def download_source_provenance(self, source_id):
+        """
+        The current Old CantusDB API endpoints do not provide
+        a method for obtaining Provenance metadata (the text returned
+        by the API in the provenance field contains more extensive notes
+        on the source provenance, but not the shorter designation that is
+        displayed publicly in CantusDB and on Cantus Ultimus). As such, this
+        method extracts this source provenance from CantusDB html. Incorporating
+        this into CantusDB's API is an open issue 
+        (see https://github.com/DDMAL/CantusDB/issues/564), so this will be
+        simplified once that is complete.
+        """
+        source_pg_url = self.cdb_base_url + "/source/" + source_id
+        source_pg_response = request.urlopen(source_pg_url).read().decode()
+        #import pdb; pdb.set_trace()
+        source_pg_html = ProvenanceParser()
+        source_pg_html.feed(source_pg_response)
+        return source_pg_html.provenance
+
+    def extract_source_data(self, source_json):
+        """
+        Extracts required elements for the CU Manuscript
+        model from the API response json.
+
+        :param json source_json: parsed json response from json-node api
+        :return: a dictionary of Manuscript attributes
+        """
+        source_dict = {}
+        source_dict["id"] = source_json["vid"]
+        source_dict["name"] = source_json["title"]
+        if source_json["field_siglum"]:
+            source_dict["siglum"] = source_json["field_siglum"]["und"][0]["value"]
+        else:
+            source_dict["siglum"] = ""
+        if source_json["field_date"]:
+            source_dict["date"] = source_json["field_date"]["und"][0]["value"]
+        else:
+            source_dict["date"] = ""
+        ## See documentation for 
+        # if source_json["field_provenance"]:
+        #     source_dict["provenance"] = source_json["field_provenance"]["und"][0]["value"]
+        # else:
+        #     source_dict["provenance"] = ""
+        source_dict["provenance"] = self.download_source_provenance(source_dict["id"])
+        if source_json["field_summary"]:
+            source_dict["description"] = source_json["field_summary"]["und"][0]["value"]
+        else:
+            source_dict["description"] = ""
+        return source_dict
+            
+    def get_source_data(self, source_id):
+        """
+        Given a CantusDB, returns a dictionary of source metadata 
```

Comment:
Should this read "Given a Source ID from CantusDB, returns..."?

Reply:
Yes, will change